File: itmo-2s-scala-hw3/src/main/scala/bank/AccountImpl.scala
package bank

import java.time.format.DateTimeFormatter._
import java.time.{Clock, LocalDateTime}
import scala.annotation.tailrec

class AccountImpl(override val balance: Int, override val operations: List[Operation], clock: Clock)
    extends Account {

  override def topUp(amount: Int): Option[Account] = {
    Option.when(amount >= 0)(
      new AccountImpl(
        balance + amount,
        Operation(LocalDateTime.now(clock), amount) :: operations,
        clock
      )
    )
  }

  override def withdraw(amount: Int): Option[Account] = {
    Option.when(!(amount < 0) && !(amount > balance))(
      new AccountImpl(
        balance - amount,
        operations :+ Operation(LocalDateTime.now(clock), -amount),
        clock
      )
    )
  }

  override def statement: String = "Time Amount Balance" + statement("", operations, 0)

  @tailrec
  private def statement(cache: String, operations: List[Operation], balance: Int): String = {
    operations match {
      case Nil => cache
      case op :: ops =>
        val time       = op.time.format(ofPattern("dd.MM.yyyy HH:mm:ss"))
        val amount     = (if (op.amount > 0) "+" else "") + op.amount
        val newBalance = balance + op.amount
        statement(cache + s"\n$time $amount $newBalance", ops, newBalance)
    }
  }

}

File: itmo-2s-golang-hw3/pkg/bank/account_impl.go
package bank

import (
	"strings"
)

type AccountImpl struct {
	clock      Clock
	operations []Operation
	balance    int
}

func (a *AccountImpl) TopUp(amount int) (Account, bool) {
	if amount <= 0 {
		return nil, false
	}

	a.balance = a.balance + amount

	a.operations = append(a.operations, Operation{
		OpTime:   a.clock.Now(),
		OpType:   TopUpOp,
		OpAmount: amount,
		Balance:  a.balance,
	})
	return a, true
}

func (a *AccountImpl) Withdraw(amount int) (Account, bool) {
	if amount <= 0 || amount > a.balance {
		return nil, false
	} else {
		a.balance = a.balance - amount
		a.operations = append(a.operations, Operation{
			OpTime:   a.clock.Now(),
			OpType:   WithdrawOp,
			OpAmount: amount,
			Balance:  a.balance,
		})
		return a, true
	}
}

func (a *AccountImpl) Operations() []Operation {
	return a.operations
}

func (a *AccountImpl) Statement() string {
	var lines []string

	for _, operation := range a.operations {
		lines = append(lines, operation.ToString())
	}
	return strings.Join(lines, "\n")
}

func (a *AccountImpl) Balance() int {
	return a.balance
}
